_delphi_utils_python/delphi_utils/flagging/flag_io.py

```diff
@@ -0,0 +1,369 @@
+                                           Key='/'.join(key.values[0]))
+                    return pd.read_csv(data['Body'], index_col=0, parse_dates=[0], header=0)
+                raw_df = raw_dates_from_file(conn, 'raw.csv', rel_files)
+                raw_dates = raw_df.index
+                # The raw dataframe was missing dates
+                if (flag_p["df_start_date"] not in raw_dates) or (
+                        flag_p["df_end_date"] not in raw_dates):
+                    gen_dict[lag_name] = base_dict
+                else:
+                    res_files = g.query("@g[5]==@pfold")
+                    final_df = raw_dates_from_file(conn, 'wkdy_corr.csv', rel_files)
+                    base_dict.update({'input_df': final_df})
+                    if res_files.empty:
+                        gen_ar_dict[f'{lag_name}/{pfold}'] = base_dict
+                    if not ((res_files.empty) and
+                            ((flag_p["eval_start_date"] in raw_dates)
+                             and (flag_p["eval_end_date"] in raw_dates))):
+                        gen_ar_dict[f'{lag_name}/{pfold}'] = base_dict
+    return gen_dict, gen_ar_dict
+
+def log_flag(files, loc_list, logger):
+    """Log method important for visualization and Slack integration."""
+    html_link = "<https://ananya-joshi-visapp-vis-523f3g.streamlitapp.com/?params="
+    for name, flags in files.items():
+        if "flagging" in name and not flags.empty:
+            event = ""
+            p_text = ""
+            for grp, df in flags.groupby('date'):
+                dt = grp.strftime('%Y/%m/%d')
+                p_text += f"*{dt}* \n"
+                df = df.sort_values(by=['val', 'state'])
+                for i, fl_df in df.reset_index(drop=True).iterrows():
+                    if "spike" in name:
+                        event = "Large Change Spike Flag"
+                        start_link = f"{html_link}{','.join(loc_list[:3])},{dt},{fl_df.state}"
+                        p_text += f"\t{start_link}|*{fl_df.state}*>\n"
+                    elif "ar" in name:
+                        event = "Autoregressive Model Flag"
+                        start_link = f"{html_link}{','.join(loc_list)},{fl_df.state}"
+                        p_text += f"\t{start_link}|*{fl_df.state}*: pval={round(fl_df.val, 2)}>\n"
+                    if (i+1)%50==0:
+                        logger.info(event,
+                                    contd = (i+1)//50,
+                                    payload=p_text,
+                                    hits=flags.shape[0])
+                        p_text = ""
+
+def save_files(loc_list, files_dict, flag_meta, logger):
+    """
+    Save files in the proper location.
+
+    Loc list is a list of parameters useful for creating output dictionaries.
+    Files dict is a dictionary of file name and their corresponding dataframe.
+
+    Offline parameter shows where to save the data.
+    """
+    log_flag(files_dict, loc_list, logger)
```

Comment:
Hypothetically, is there ever a circumstance where we'd want to flag but not save files? Or what if file-saving fails for some reason but we still get the alert -- is that okay?